versioneer/versioneer.go

```diff
@@ -0,0 +1,260 @@
+
+	return fmt.Sprintf("%s/%s:%s", registryAndOrg, a.Flavor, tag), nil
+}
+
+func (a *Artifact) BaseContainerName(registryAndOrg, id string) (string, error) {
+	if a.Flavor == "" {
+		return "", errors.New("Flavor is empty")
+	}
+
+	if id == "" {
+		return "", errors.New("no id passed")
+	}
+
+	tag, err := a.BaseTag()
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s/%s:%s-%s", registryAndOrg, a.Flavor, tag, id), nil
+}
+
+func (a *Artifact) BaseTag() (string, error) {
+	return a.commonName()
+}
+
+func (a *Artifact) Tag() (string, error) {
+	commonName, err := a.commonVersionedName()
+	if err != nil {
+		return commonName, err
+	}
+
+	return strings.ReplaceAll(commonName, "+", "-"), nil
+}
+
+// OSReleaseVariables returns a set of variables to be appended in /etc/os-release
+func (a *Artifact) OSReleaseVariables(registryAndOrg, githubRepo, bugURL, homeURL string) (string, error) {
+	if registryAndOrg == "" {
+		return "", errors.New("registry-and-org must be set")
+	}
+	commonName, err := a.commonVersionedName()
+	if err != nil {
+		return commonName, err
+	}
+	kairosName := fmt.Sprintf("kairos-%s-%s-%s", a.Variant, a.Flavor, a.FlavorRelease)
+	kairosVersion := a.Version
+	if a.SoftwareVersion != "" {
+		kairosVersion += "-" + strings.ReplaceAll(a.SoftwareVersion, "+", "-")
+	}
+
+	containerName, err := a.ContainerName(registryAndOrg)
+	if err != nil {
+		return "", err
+	}
+
+	tag, err := a.Tag()
+	if err != nil {
+		return "", err
+	}
+
+	bootableName, err := a.BootableName()
+	if err != nil {
+		return "", err
+	}
+
+	vars := map[string]string{
+		// Legacy variables (not used by versioneer)
+		"KAIROS_NAME":        kairosName,
+		"KAIROS_VERSION":     kairosVersion,
+		"KAIROS_ID":          "kairos",
+		"KAIROS_ID_LIKE":     kairosName,
+		"KAIROS_VERSION_ID":  kairosVersion,
+		"KAIROS_PRETTY_NAME": fmt.Sprintf("%s %s", kairosName, kairosVersion),
+		"KAIROS_IMAGE_REPO":  containerName,
+		"KAIROS_IMAGE_LABEL": tag,
+		"KAIROS_ARTIFACT":    bootableName,
+		// Actively used variables
+		"KAIROS_FLAVOR":         a.Flavor,
+		"KAIROS_FLAVOR_RELEASE": a.FlavorRelease,
+		"KAIROS_VARIANT":        a.Variant,
+		"KAIROS_MODEL":          a.Model,
+		"KAIROS_ARCH":           a.Arch,
+		"KAIROS_RELEASE":        a.Version,
+	}
+	if bugURL != "" {
+		vars["KAIROS_BUG_REPORT_URL"] = bugURL
+	}
+	if homeURL != "" {
+		vars["KAIROS_HOME_URL"] = homeURL
+	}
+	if githubRepo != "" {
+		vars["KAIROS_GITHUB_REPO"] = githubRepo
+	}
+	if a.SoftwareVersion != "" {
+		vars["KAIROS_SOFTWARE_VERSION"] = a.SoftwareVersion
+	}
+
+	result := ""
+	for k, v := range vars {
+		result += fmt.Sprintf("%s=%s\n", k, v)
+	}
+
+	return result, nil
+}
+
+func (a *Artifact) TagList(registryAndOrg string) (TagList, error) {
+	if a.RegistryInspector == nil {
+		a.RegistryInspector = &DefaultRegistryInspector{}
+	}
+
+	return a.RegistryInspector.TagList(registryAndOrg, a)
+}
+
+func (a *Artifact) commonName() (string, error) {
+	if err := a.Validate(); err != nil {
```

Comment:
we only validate here, should we do it in a lot more places?

Reply:
All the others are eventually calling this one. For some of them, I've added a test to make sure it is indeed validated. E.g. https://github.com/kairos-io/kairos-sdk/pull/59/files/d5a87ece6e62a02b547bac516ce1a0ba5b6e5183#diff-f8a8b10ab615219af502040ef4bfc01f8e25100aa2adf76de32a88a0277ccb6bR54

I probably didn't test that for all methods calling `commonName()` but all of them do.